### PR TITLE
Simplify librustc_errors

### PR DESCRIPTION
**Pull request by sophiajt**
This is part 2 of the error crate refactor, starting with #34403.

In this refactor, I focused on slimming down the error crate to fewer moving parts.  As such, I've removed quite a few parts and replaced the with simpler, straight-line code.  Specifically, this PR:
- Removes BasicEmitter
- Remove emit from emitter, leaving emit_struct
- Renames emit_struct to emit
- Removes CoreEmitter and focuses on a single Emitter
- Implements the latest changes to error format RFC (#1644)
- Removes (now-unused) code in emitter.rs and snippet.rs
- Moves more tests to the UI tester, removing some duplicate tests in the process

There is probably more that could be done with some additional refactoring, but this felt like it was getting to a good state.

r? @alexcrichton   cc: @Manishearth (as there may be breaking changes in stuff I removed/changed)

### CODE REVIEW

File: src/librustc_errors/emitter.rs [hunk 1]
```diff
@@ -10,89 +10,39 @@
...
-        self.emit_message(&FullSpan(msp.clone()),
-                          msg,
-                          code,
-                          lvl,
-                          true,
-                          true);
-    }
+impl Emitter for EmitterWriter {
+    fn emit(&mut self, db: &DiagnosticBuilder) {
+        let old_school = match self.format_mode {
```
Comment (alexcrichton) [inline src/librustc_errors/emitter.rs]: I was massively confused about "school" here vs "skool", please udpate.

Comment (alexcrichton) [inline src/librustc_errors/emitter.rs]: To be clear, I think "skool" is the correct spelling

**Comment by alexcrichton** [comment]
Looks good to me! I didn't look too closely at the code changes as it looked like it was mostly just stuff moving around, but I like the simplifying refactors here quite a bit!

r? @Manishearth -- if this plays into a syntax breaking change I'll defer to you, otherwise this looks ready to r+ to me

File: src/librustc_errors/emitter.rs (hunk 1 above)
Comment (jonathandturner) [inline src/librustc_errors/emitter.rs]: -_-

**Comment by Manishearth** [comment]
Not really a libsyntax breaking change, but might affect syntex -- cc @dtolnay

**Comment by jonathandturner** [comment]
Looks like we got the okay from @Manishearth and @dtolnay.

@bors r=alexcrichton

**Comment by alexcrichton** [comment]
@bors: retry

* Maybe https://github.com/rust-lang/rust/issues/33361

**Pull request merged**